bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -673,6 +686,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
   }
 }
 
+bool ReplicaImp::validatePreProcessedResultSignatures(preprocessor::PreProcessedResultMsg &msg) {
+  auto [buf, buf_len] = msg.getResultSignaturesBuf();
+  auto sigs = preprocessor::PreProcessResultSignature::deserializeResultSignatureList(buf, buf_len);
+
+  // f because the primary also executes the request (check PreProcessor::numOfRequiredReplies())
+  const auto expectedSignatureCount = getReplicaConfig().fVal;
+  if (sigs.size() < expectedSignatureCount) {
+    LOG_ERROR(GL,
+              "PreProcessedResult signatures validation failure - not enough signatures. Expected "
+                  << expectedSignatureCount << " got " << sigs.size());
+    return false;
+  }
+
+  for (const auto &s : sigs) {
+    auto hash = concord::util::SHA3_256().digest(msg.requestBuf(), msg.requestLength());
```

Comment:
We should create the hashes in parallel. Its done in PrePrepareMsg creation.

Reply:
Good point, I'll do it.

Reply:
Actually the hash is calculated only once and should be outside the for loop. Good catch!